pkg/controller/ec2/instance/controller.go

```diff
@@ -221,6 +176,86 @@ func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.E
 	}, nil
 }
 
+func (e *external) describeInstance(ctx context.Context, instanceId string) (
+	*types.Instance,
+	awsec2.DescribeInstanceAttributeOutput,
+	error,
+) {
+	wg := sync.WaitGroup{}
+
+	var describeOutput *awsec2.DescribeInstancesOutput
+	var describeError error
+	wg.Add(1)
+	go func() {
+		describeOutput, describeError = e.client.DescribeInstances(ctx, &awsec2.DescribeInstancesInput{
+			InstanceIds: []string{instanceId},
+		})
+		wg.Done()
+	}()
+
+	attrs := awsec2.DescribeInstanceAttributeOutput{}
+	attrsErr := atomic.Pointer[error]{}
+	descAttr := func(attr types.InstanceAttributeName) (*awsec2.DescribeInstanceAttributeOutput, error) {
+		return e.client.DescribeInstanceAttribute(ctx, &awsec2.DescribeInstanceAttributeInput{
+			InstanceId: &instanceId,
+			Attribute:  attr,
+		})
+	}
```

Comment:
I wonder if this could be done more elegantly by iterating over an array of attribute keys and start a Go routine in parallel. I would also prefer to use https://pkg.go.dev/golang.org/x/sync/errgroup instead of standard `WaitGroup`.

Something like this:

```go
func (e *external) describeInstanceAttributes(ctx context.Context, cr *svcapitypes.Instance) (*awsec2.DescribeInstanceAttributeOutput, error) {
	eg := errgroup.Group{}
	out := &awsec2.DescribeInstanceAttributeOutput{}
	describeAttribute := func(attr types.InstanceAttributeName, onSuccess func(res *awsec2.DescribeInstanceAttributeOutput)) {
		eg.Go(func() error {
			res, err := e.client.DescribeInstanceAttribute(ctx, &awsec2.DescribeInstanceAttributeInput{
				InstanceId: aws.String(meta.GetExternalName(cr)),
				Attribute: attr,
			})
			if err != nil {
				err = errors.Wrap(err, string(attr))
				return err
			}
			onSuccess(res)
			return nil
		})
	}

	describeAttribute(types.InstanceAttributeNameDisableApiTermination, func(res *awsec2.DescribeInstanceAttributeOutput) {
		out.DisableApiTermination = res.DisableApiTermination
	})
	describeAttribute(types.InstanceAttributeNameInstanceInitiatedShutdownBehavior, func(res *awsec2.DescribeInstanceAttributeOutput) {
		out.InstanceInitiatedShutdownBehavior = res.InstanceInitiatedShutdownBehavior
	})
	describeAttribute(types.InstanceAttributeNameUserData, func(res *awsec2.DescribeInstanceAttributeOutput) {
		out.UserData = res.UserData
	})
	
	err := eg.Wait()
	return out, err
}
```

Reply:
Oh, cool. I didn't know about `errgroup`. It looks much better now.